File: JSReact/learning_react_app/src/Shape.js
// условный рендеринг по состоянию

import { useEffect, useState } from "react";



const Shape = () => {
    const [shape, setShape] = useState(null);
    const [randomNumber, setRandomNumber] = useState(null);
    useEffect(() => {
        const number = Math.floor(Math.random() * 100);
        setRandomNumber(number);
        if (randomNumber > 5) {
            setShape('circle');
        } else {
            setShape('sqrt');
        }
    }, []);

    return (
        <div>
            {shape === "sqrt" ?
             <div style={{
                width: "200px",
                height: "200px",
                background: "blue",
                color: "white",
                display: "flex",
                justifyContent: "center",
                alignItems: "center",
                }}>
                {randomNumber}
             </div> :
             <div style={{
                width: "200px",
                borderRadius: "50%",
                height: "200px",
                background: "red",
                color: "white",
                display: "flex",
                justifyContent: "center",
                alignItems: "center",
                }}>
                {randomNumber}
             </div>}
        </div>
    );
}

export default Shape;

File: Java_2/3_SOLID/src/main/java/org/example/2_O/Shape.java
// 2. Принцип открытости/закрытости (Open/Closed Principle - OCP):
/*
public class Circle {
    public double radius;
    public double calculateArea() {
        return Math.PI * radius * radius;
    }
}
public class Rectangle {
    public double width;
    public double height;
    public double calculateArea() {
        return width * height;
    }
}
*/
// В этом примере для добавления новых фигур (например, треугольников) нам пришлось бы изменять существующий код.
// Рефакторинг:
public interface Shape {
    double calculateArea();
}
public class Circle implements Shape {
    private double radius;
    @Override
    public double calculateArea() {
        return Math.PI * radius * radius;
    }
}
public class Rectangle implements Shape {
    private double width;
    private double height;
    @Override
    public double calculateArea() {
        return width * height;
    }
}
// Теперь мы можем легко добавлять новые фигуры, не изменяя существующий код